src/gui/title_bar.cpp

```diff
@@ -0,0 +1,600 @@
+	}
+}
+
+// ***************************************************************************
+// External notifications and setter functions
+// ***************************************************************************
+
+void GFX_NotifyBooting()
+{
+	state.guest_os_booted = true;
+	GFX_RefreshTitle();
+}
+
+void GFX_SetAudioCaptureMark(const bool is_capturing)
+{
+	if (state.audio_capture != is_capturing) {
+		state.audio_capture = is_capturing;
+		GFX_RefreshTitle();
+	}
+}
+
+void GFX_SetVideoCaptureMark(const bool is_capturing)
+{
+	if (state.video_capture != is_capturing) {
+		state.video_capture = is_capturing;
+		GFX_RefreshTitle();
+	}
+}
+
+void GFX_SetProgramName(const std::string& segment_name,
+                        const std::string& canonical_name)
+{
+	std::string segment_name_dos = segment_name;
+
+	// Segment name might contain just about any character - adapt it
+	for (size_t i = 0; i < segment_name_dos.size(); ++i) {
+		if (!is_extended_printable_ascii(segment_name_dos[i])) {
+			segment_name_dos[i] = '?';
+		}
+	}
+	trim(segment_name_dos);
+
+	// Store new names as UTF-8, refresh title bar
+	dos_to_utf8(segment_name_dos, state.segment_name);
+	dos_to_utf8(canonical_name, state.canonical_name);
+	GFX_RefreshTitle();
+}
+
+void GFX_SetCycles(const int32_t cycles)
+{
+	if (cycles >= 0 && state.num_cycles != cycles) {
+		state.num_cycles = cycles;
+		GFX_RefreshTitle();
+	}
+}
+
+void GFX_SetMouseHint(const MouseHint hint_id)
+{
+	static const std::string prexix = " - ";
+
+	auto create_hint_str = [](const char* requested_name) {
+		char hint_buffer[200] = {0};
+
+		safe_sprintf(hint_buffer, MSG_GetRaw(requested_name), PRIMARY_MOD_NAME);
+		return prexix + hint_buffer;
+	};
+
+	std::string hint = {};
+	switch (hint_id) {
+	case MouseHint::None:
+		break;
+	case MouseHint::CapturedHotkey:
+		hint = create_hint_str("TITLEBAR_HINT_CAPTURED_HOTKEY");
+		break;
+	case MouseHint::CapturedHotkeyMiddle:
+		hint = create_hint_str("TITLEBAR_HINT_CAPTURED_HOTKEY_MIDDLE");
+		break;
+	case MouseHint::ReleasedHotkey:
+		hint = create_hint_str("TITLEBAR_HINT_RELEASED_HOTKEY");
+		break;
+	case MouseHint::ReleasedHotkeyMiddle:
+		hint = create_hint_str("TITLEBAR_HINT_RELEASED_HOTKEY_MIDDLE");
+		break;
+	case MouseHint::ReleasedHotkeyAnyButton:
+		hint = create_hint_str("TITLEBAR_HINT_RELEASED_HOTKEY_ANY_BUTTON");
+		break;
+	case MouseHint::SeamlessHotkey:
+		hint = create_hint_str("TITLEBAR_HINT_SEAMLESS_HOTKEY");
+		break;
+	case MouseHint::SeamlessHotkeyMiddle:
+		hint = create_hint_str("TITLEBAR_HINT_SEAMLESS_HOTKEY_MIDDLE");
+		break;
+	default: assert(false); break;
+	}
+
+	if (state.hint_mouse != hint) {
+		state.hint_mouse = std::move(hint);
+		GFX_RefreshTitle();
+	}
+}
+
+// ***************************************************************************
+// Lifecycle
+// ***************************************************************************
+
+void TITLEBAR_ReadConfig(Section* section)
```

Comment:
If you'd return a struct with the configured options (`std::optional<TitlebarConfig>`), then this function would become testable. I recommend doing so.

Reply:
OK, submodule interface changed to be testable.